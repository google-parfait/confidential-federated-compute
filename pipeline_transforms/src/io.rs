// Copyright 2023 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

extern crate alloc;

use crate::proto::{
    record::{HpkePlusAeadData, Kind as RecordKind},
    Record,
};
use alloc::vec::Vec;
use anyhow::anyhow;
use bitmask::bitmask;

bitmask! {
    /// The set of decryption modes supported by a RecordDecoder.
    pub mask DecryptionModeSet : u8 where
    /// A specific decryption mode, corresponding to an `Input.kind`.
    flags DecryptionMode {
        Unencrypted = 1,
        HpkePlusAead = 2,
    }
}

/// Decodes pipeline_transforms Record messages.
pub struct RecordDecoder {
    allowed_modes: DecryptionModeSet,
    private_key: cfc_crypto::PrivateKey,
    public_key: Vec<u8>,
}

impl RecordDecoder {
    /// Constructs a new RecordDecoder.
    pub fn new(allowed_modes: DecryptionModeSet) -> Self {
        let (private_key, public_key) = cfc_crypto::gen_keypair();
        Self {
            allowed_modes,
            private_key,
            public_key,
        }
    }

    pub fn public_key(&self) -> &[u8] {
        &self.public_key
    }

    /// Decodes an `Record` message to the corresponding message bytes.
    ///
    /// # Return Value
    ///
    /// Returns the unencrypted message on success.
    pub fn decode(&self, record: &Record) -> anyhow::Result<Vec<u8>> {
        match &record.kind {
            Some(RecordKind::UnencryptedData(d)) => {
                if !self.allowed_modes.contains(DecryptionMode::Unencrypted) {
                    return Err(anyhow!("Record.unencrypted_data is not supported"));
                }
                Ok(d.clone())
            }

            Some(RecordKind::HpkePlusAeadData(msg)) => {
                if !self.allowed_modes.contains(DecryptionMode::HpkePlusAead) {
                    return Err(anyhow!("Record.hpke_plus_aead_data is not supported"));
                }
                cfc_crypto::decrypt_message(
                    &msg.ciphertext,
                    &msg.ciphertext_associated_data,
                    &msg.encrypted_symmetric_key,
                    msg.encrypted_symmetric_key_associated_data
                        .as_ref()
                        .unwrap_or(&msg.ciphertext_associated_data),
                    &msg.encapsulated_public_key,
                    &self.private_key,
                )
            }

            _ => Err(anyhow!("unsupported Record kind")),
        }
    }
}

impl Default for RecordDecoder {
    /// Creates an `RecordDecoder` that supports all decryption modes.
    fn default() -> Self {
        Self::new(DecryptionModeSet::all())
    }
}

/// The type of encryption that should be performed by a RecordEncoder.
#[derive(Clone, Copy)]
pub enum EncryptionMode<'a> {
    Unencrypted,
    HpkePlusAead {
        public_key: &'a [u8],
        associated_data: &'a [u8],
    },
}

/// Encodes pipeline_transforms Record messages.
#[derive(Default)]
pub struct RecordEncoder;

impl RecordEncoder {
    /// Constructs a new `RecordEncoder`.
    pub fn new() -> Self {
        Self::default()
    }

    /// Encodes data as a `Record` message.
    pub fn encode(&self, mode: EncryptionMode, data: &[u8]) -> anyhow::Result<Record> {
        let kind = match mode {
            EncryptionMode::Unencrypted => RecordKind::UnencryptedData(data.to_vec()),

            EncryptionMode::HpkePlusAead {
                public_key,
                associated_data,
            } => {
                let (ciphertext, encapsulated_public_key, encrypted_symmetric_key) =
                    cfc_crypto::encrypt_message(data, public_key, associated_data)?;
                RecordKind::HpkePlusAeadData(HpkePlusAeadData {
                    ciphertext,
                    ciphertext_associated_data: associated_data.to_vec(),
                    encrypted_symmetric_key,
                    encapsulated_public_key,
                    ..Default::default()
                })
            }
        };
        Ok(Record { kind: Some(kind) })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_unencrypted() -> anyhow::Result<()> {
        let input = Record {
            kind: Some(RecordKind::UnencryptedData(b"data".to_vec())),
            ..Default::default()
        };

        let decoder = RecordDecoder::new(DecryptionMode::Unencrypted.into());
        assert_eq!(decoder.decode(&input)?, b"data".to_vec());
        Ok(())
    }

    #[test]
    fn test_decode_unencrypted_disallowed() {
        let input = Record {
            kind: Some(RecordKind::UnencryptedData(b"data".to_vec())),
            ..Default::default()
        };

        let decoder = RecordDecoder::new(DecryptionModeSet::none());
        assert!(decoder.decode(&input).is_err());
    }

    #[test]
    fn test_decode_hpke_plus_aead() -> anyhow::Result<()> {
        let plaintext = b"data";
        let ciphertext_associated_data = b"associated data".to_vec();

        let decoder = RecordDecoder::new(DecryptionMode::HpkePlusAead.into());
        let (ciphertext, encapsulated_public_key, encrypted_symmetric_key) =
            cfc_crypto::encrypt_message(
                plaintext,
                decoder.public_key(),
                &ciphertext_associated_data,
            )?;
        let input = Record {
            kind: Some(RecordKind::HpkePlusAeadData(HpkePlusAeadData {
                ciphertext,
                ciphertext_associated_data,
                encrypted_symmetric_key,
                encapsulated_public_key,
                ..Default::default()
            })),
            ..Default::default()
        };
        assert_eq!(decoder.decode(&input)?, plaintext.to_vec());
        Ok(())
    }

    #[test]
    fn test_decode_hpke_plus_aead_disallowed() -> anyhow::Result<()> {
        let plaintext = b"data";
        let ciphertext_associated_data = b"associated_data".to_vec();

        let decoder = RecordDecoder::new(DecryptionModeSet::none());
        let (ciphertext, encapsulated_public_key, encrypted_symmetric_key) =
            cfc_crypto::encrypt_message(
                plaintext,
                decoder.public_key(),
                &ciphertext_associated_data,
            )?;
        let input = Record {
            kind: Some(RecordKind::HpkePlusAeadData(HpkePlusAeadData {
                ciphertext,
                ciphertext_associated_data,
                encrypted_symmetric_key,
                encapsulated_public_key,
                ..Default::default()
            })),
            ..Default::default()
        };
        assert!(decoder.decode(&input).is_err());
        Ok(())
    }

    #[test]
    fn test_decode_hpke_plus_aead_with_decryption_error() -> anyhow::Result<()> {
        let plaintext = b"data";
        let ciphertext_associated_data = b"associated_data".to_vec();

        let decoder = RecordDecoder::new(DecryptionMode::HpkePlusAead.into());
        let (ciphertext, encapsulated_public_key, encrypted_symmetric_key) =
            cfc_crypto::encrypt_message(
                plaintext,
                decoder.public_key(),
                &ciphertext_associated_data,
            )?;
        let input = Record {
            kind: Some(RecordKind::HpkePlusAeadData(HpkePlusAeadData {
                ciphertext,
                ciphertext_associated_data: b"wrong associated data".to_vec(),
                encrypted_symmetric_key,
                encapsulated_public_key,
                ..Default::default()
            })),
            ..Default::default()
        };
        assert!(decoder.decode(&input).is_err());
        Ok(())
    }

    #[test]
    fn test_encode_unencrypted() -> anyhow::Result<()> {
        let data = b"data";
        assert_eq!(
            RecordEncoder::default().encode(EncryptionMode::Unencrypted, data)?,
            Record {
                kind: Some(RecordKind::UnencryptedData(data.to_vec())),
                ..Default::default()
            }
        );
        Ok(())
    }

    #[test]
    fn test_encode_hpke_plus_aead() -> anyhow::Result<()> {
        let plaintext = b"data";
        let associated_data = b"associated data";
        let (private_key, public_key) = cfc_crypto::gen_keypair();
        let output = RecordEncoder::default().encode(
            EncryptionMode::HpkePlusAead {
                public_key: &public_key,
                associated_data,
            },
            plaintext,
        )?;
        let msg = match output.kind {
            Some(RecordKind::HpkePlusAeadData(msg)) => msg,
            _ => return Err(anyhow!("expected HpkePlusAeadData")),
        };

        assert_eq!(&msg.ciphertext_associated_data, associated_data);
        // encrypted_symmetric_key_associated_data should not be set.
        assert_eq!(msg.encrypted_symmetric_key_associated_data, None);
        assert_eq!(
            cfc_crypto::decrypt_message(
                &msg.ciphertext,
                &msg.ciphertext_associated_data,
                &msg.encrypted_symmetric_key,
                &msg.ciphertext_associated_data,
                &msg.encapsulated_public_key,
                &private_key
            )?,
            plaintext
        );
        Ok(())
    }

    #[test]
    fn test_encode_hpke_plus_aead_with_encryption_error() {
        let plaintext = b"data";
        let associated_data = b"associated data";
        assert!(RecordEncoder::default()
            .encode(
                EncryptionMode::HpkePlusAead {
                    public_key: b"invalid key",
                    associated_data,
                },
                plaintext,
            )
            .is_err());
    }
}
