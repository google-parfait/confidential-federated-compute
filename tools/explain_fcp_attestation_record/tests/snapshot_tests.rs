// Copyright 2024 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Tests the 'explain_fcp_attestation_record' binary by running it over some
//! checked-in record data and comparing its output against snapshot files
//! containing the expected output. We use this testing approach because the
//! tool output generally quite long, and because we don't fully control the
//! format of it (the Oak Attestation Explanation library produces a good chunk
//! of the output), and hence it's easier to simply maintain and update 'golden'
//! outputs as needed.
//!
//! To make updating and reviewing changes to the snapshot files easier you can
//! use the `cargo insta review` command (installable using `cargo install
//! insta`). The test will generate updated snapshots with a '.new' filename
//! suffix whenever there's a test failure, however, and you can just manually
//! move those files into the original snapshot files, without the use of the
//! 'cargo insta' tool.

use std::io::Write;

use assert_cmd::Command;
use insta::assert_snapshot;
use prost::Message as _;

mod testdata;

/// Runs the explain tool over a record file that only contains valid ledger
/// attestation evidence and an empty access policy (which is not particularly
/// useful in practice, since it prevents any form of further data processing).
#[test]
fn test_explain_record_with_empty_data_access_policy_from_file() -> anyhow::Result<()> {
    // Create a temporary file containing the
    let mut record_file = tempfile::NamedTempFile::new()?;
    let serialized_record = testdata::record_with_empty_access_policy().encode_to_vec();
    record_file.write_all(&serialized_record)?;
    let record_file_path = record_file.path().to_str().unwrap();

    // Note: we run this test by actually running the command line binary. This
    // helps ensure that the binary actually works end-to-end, including
    // argument parsing etc. Subsequent tests forego this step and instead call
    // directly into the helper functions upon which the library is built.
    let cmd = Command::new(std::env::var("EXPLAIN_TOOL").unwrap())
        .arg("--record")
        .arg(record_file_path)
        .assert()
        .success();
    let output = std::str::from_utf8(&cmd.get_output().stdout)?;

    // Check that the expected output in the snapshot matches the actual output in
    // the buffer.
    assert_snapshot!(output.replace(record_file_path, "{TMP_RECORD_FILE}"));
    Ok(())
}

/// Like [`test_explain_record_with_empty_data_access_policy_from_file`] but
/// passes the record via stdin.
#[test]
fn test_explain_record_with_empty_data_access_policy_from_stdin() -> anyhow::Result<()> {
    let serialized_record = testdata::record_with_empty_access_policy().encode_to_vec();

    // Note: we run this test by actually running the command line binary. This
    // helps ensure that the binary actually works end-to-end, including
    // argument parsing etc. Subsequent tests forego this step and instead call
    // directly into the helper functions upon which the library is built.
    let cmd = Command::new(std::env::var("EXPLAIN_TOOL").unwrap())
        .arg("--record")
        .arg("-")
        .write_stdin(serialized_record)
        .assert()
        .success();
    let output = std::str::from_utf8(&cmd.get_output().stdout)?;

    // Check that the expected output in the snapshot matches the actual output in
    // the buffer.
    assert_snapshot!(output);
    Ok(())
}

/// Runs the explain tool over a record that contains valid ledger attestation
/// evidence and an access policy that is populated with a few transforms and
/// access budgets.
#[test]
fn test_explain_record_with_nonempty_data_access_policy() -> anyhow::Result<()> {
    // Note: we don't actually invoke the binary, and instead call the
    // explain_fcp_attestation_record library upon which the binary is built. This
    // makes it a bit easier to pass in records which we constructed/modified
    // within the test code, without having to write to temporary files etc.
    let mut buf = String::new();
    explain_fcp_attestation_record::explain_record(
        &mut buf,
        &testdata::record_with_nonempty_access_policy(),
    )?;

    // Check that the expected output in the snapshot matches the actual output in
    // the buffer.
    assert_snapshot!(buf);
    Ok(())
}
