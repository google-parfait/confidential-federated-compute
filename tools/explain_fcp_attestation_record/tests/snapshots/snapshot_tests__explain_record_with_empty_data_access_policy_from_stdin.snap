---
source: tools/explain_fcp_attestation_record/tests/snapshot_tests.rs
expression: output
---
Inspecting AttestationVerificationRecord provided via stdin.

========================================
===== LEDGER ATTESTATION EVIDENCE ======
========================================

Oak Restricted Kernel Stack in a AMD SEV-SNP TEE

_____ Root Layer _____

The attestation is rooted in an AMD SEV-SNP TEE.

Attestations identifying the firmware captured in the evidence can be found here:
https://search.sigstore.dev/?hash=a8c51290169976afc37e6e6d866107285b5f4711a9ce5389c05d9a5d297d68c5

â“˜ The firmware attestation digest is the SHA2-256 hash of the SHA2-384 hash of the initial memory state taken by the AMD SoC. The original SHA2-384 hash of the initial memory is: SHA2-384:571e632335e16997bdc312208d540b083518fe05d84b8954a6529a019a04229f25347fdaac1bc80418addc64a3d48704; it is listed as the 'initial_measurement' in the evidence of this layer.

The evidence describing this layer is outlined below.

sev_snp:
  current_tcb:
    boot_loader: 3
    fmc: 0
    microcode: 209
    snp: 22
    tee: 0
  debug: false
  hardware_id: c1f5c58f728e2eded313ee675ac982393169c9bbffe6e250c54d81fcea2dad556d314f3c62c16cb9ae3d22f68a747261835f220d3656d0e6b646e6b2f11252cf
  initial_measurement: 571e632335e16997bdc312208d540b083518fe05d84b8954a6529a019a04229f25347fdaac1bc80418addc64a3d48704
  product: 2
  report_data: f75ac8ac3bfd479a0f121c384d71b4c032867895db71d91d135710083fd780a10000000000000000000000000000000000000000000000000000000000000000
  reported_tcb:
    boot_loader: 3
    fmc: 0
    microcode: 209
    snp: 22
    tee: 0
  vmpl: 0


Note: binaries for this layer are generally provided by the Oak project (https://github.com/project-oak/oak).

_____ Kernel Layer _____

Attestations identifying the binaries captured in the evidence in this layer can be found as outlined below.
Kernel: https://search.sigstore.dev/?hash=4b1e70ad0ad326f3ee6f8f45f77358f0b8bb5df05321a5abc34c66022e27450b
Initial Ramdisk: https://search.sigstore.dev/?hash=51534334403d87176dc66406a07b5108d51f46a8534497c21f2769d1217c51b8

The evidence describing the kernel layer is outlined below.

acpi:
  sha2_256: dbaccae7bfbf006e2b8623a82f1a5fcda2ea0392233c26b18356b3bcfac231eb
init_ram_fs:
  sha2_256: 51534334403d87176dc66406a07b5108d51f46a8534497c21f2769d1217c51b8
kernel_image:
  sha2_256: 4b1e70ad0ad326f3ee6f8f45f77358f0b8bb5df05321a5abc34c66022e27450b
kernel_raw_cmd_line: console=ttyS0
kernel_setup_data:
  sha2_256: 4cd020820da663063f4185ca14a7e803cd7c9ca1483c64e836db840604b6fac1
memory_map:
  sha2_256: 807625f963ce93abe1f2d9965b5da03c66d476aec9756fded405ce5f41182b6f


Note: binaries for this layer are generally provided by the Oak project (https://github.com/project-oak/oak).

_____ Application Layer _____

The evidence describing the application is outlined below.

binary:
  sha2_256: 5d10d8013345814e07141c6a4c9297d37653239132749574a2a71483c413e9fe
config: {}


Note: this layer describes the "ledger" application binary, which is generally a build of the `ledger_enclave_app` in the https://github.com/google-parfait/confidential-federated-compute repository.


========================================
========== DATA ACCESS POLICY ==========
========================================

The data access policy allows 0 data transformations and defines 0 shared access budgets.
